planner/core/preprocess.go

```diff
@@ -1456,9 +1459,31 @@ func (p *preprocessor) checkFuncCastExpr(node *ast.FuncCastExpr) {
 }
 
 // handleAsOfAndReadTS tries to handle as of closure, or possibly read_ts.
-// If read_ts is not nil, it will be consumed.
-// If as of is not nil, timestamp is used to get the history infoschema from the infocache.
 func (p *preprocessor) handleAsOfAndReadTS(node *ast.AsOfClause) {
+	// When statement is during the Txn, we check whether there exists AsOfClause. If exists, we will return error,
+	// otherwise we should directly set the return param from TxnCtx.
+	if p.ctx.GetSessionVars().InTxn() {
+		if node != nil {
+			p.err = ErrAsOf.FastGenWithCause("as of timestamp can't be set in transaction.")
+			return
+		}
+		if p.initedLastSnapshotTS {
```

Comment:
This line is wrong. This check is not only for init, but also check for different tso. It should not early return.

Reply:
updated.